File: frontend/src/app/codezen/service/codezen.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable } from 'rxjs';
import { ProjectRequest } from '../../models/project-request';
import { ProjectResponse } from '../../models/project-response';
import { ReviewRequest } from '../../models/review-request';
import { ReviewResponse } from '../../models/review-response';
import { GuidelineRequest } from '../../models/guideline-request';
import { GuidelineResponse } from '../../models/guideline-response';
import { CommentRequest } from '../../models/comment-request';
import { CommentResponse } from '../../models/comment-response';
import {StorageService} from '../../services/storage/storage.service';

/**
 * CodeZenService - Handles all communication with CodeZen backend API.
 * Manages projects, reviews, and guidelines.
 */
@Injectable({
  providedIn: 'root'
})
export class CodezenService {
  private baseUrl = 'http://localhost:8080/api/v1/projects';

  constructor(private http: HttpClient) { }

  /**
   * Get authorization headers with JWT token
   */
  private getAuthHeaders(): HttpHeaders {
    const token = StorageService.getToken();
    if (!token) {
      console.error('No token found in storage');
      return new HttpHeaders();
    }
    return new HttpHeaders().set('Authorization', `Bearer ${token}`);
  }
  /**
   * Create a new project
   */
  createProject(request: ProjectRequest): Observable<ProjectResponse> {
    return this.http.post<ProjectResponse>(
      this.baseUrl,
      request,
      { headers: this.getAuthHeaders() }
    );
  }

  /**
   * Get all projects for the current user
   */
  getAllProjects(): Observable<ProjectResponse[]> {
    return this.http.get<ProjectResponse[]>(
      this.baseUrl,
      { headers: this.getAuthHeaders() }
    );
  }

  /**
   * Get a specific project by ID
   */
  getProject(id: number): Observable<ProjectResponse> {
    return this.http.get<ProjectResponse>(
      `${this.baseUrl}/${id}`,
      { headers: this.getAuthHeaders() }
    );
  }

  /**
   * Delete a project
   */
  deleteProject(id: number): Observable<void> {
    return this.http.delete<void>(
      `${this.baseUrl}/${id}`,
      { headers: this.getAuthHeaders() }
    );
  }

  /**
   * Submit code for review
   */
  createReview(projectId: number, request: ReviewRequest): Observable<ReviewResponse> {
    return this.http.post<ReviewResponse>(
      `${this.baseUrl}/${projectId}/reviews`,
      request,
      { headers: this.getAuthHeaders() }
    );
  }

  /**
   * Get all reviews for a project
   */
  getReviews(projectId: number): Observable<ReviewResponse[]> {
    return this.http.get<ReviewResponse[]>(
      `${this.baseUrl}/${projectId}/reviews`,
      { headers: this.getAuthHeaders() }
    );
  }

  /**
   * Get a specific review by ID
   */
  getReview(projectId: number, reviewId: number): Observable<ReviewResponse> {
    return this.http.get<ReviewResponse>(
      `${this.baseUrl}/${projectId}/reviews/${reviewId}`,
      { headers: this.getAuthHeaders() }
    );
  }

  /**
   * Add a custom guideline to a project
   */
  addGuideline(projectId: number, request: GuidelineRequest): Observable<GuidelineResponse> {
    return this.http.post<GuidelineResponse>(
      `${this.baseUrl}/${projectId}/guidelines`,
      request,
      { headers: this.getAuthHeaders() }
    );
  }

  /**
   * Get all guidelines for a project
   */
  getGuidelines(projectId: number): Observable<GuidelineResponse[]> {
    return this.http.get<GuidelineResponse[]>(
      `${this.baseUrl}/${projectId}/guidelines`,
      { headers: this.getAuthHeaders() }
    );
  }

  /**
   * Parse LLM response from JSON string to structured object
   * Handles cases where Ollama returns text instead of pure JSON
   */
  parseLLMResponse(llmResponse: string): any {
    if (!llmResponse || !llmResponse.trim()) {
      return {
        summary: 'No review response received',
        findings: [],
        effortEstimation: 'N/A'
      };
    }

    try {
      // Try to parse as direct JSON first
      const directParse = JSON.parse(llmResponse.trim());
      if (directParse && typeof directParse === 'object') {
        return this.validateAndEnhanceResponse(directParse);
      }
    } catch (e) {
      // Not direct JSON, continue to extraction
    }

    try {
      // Extract JSON object from response (may be wrapped in markdown or text)
      // Match first { ... } that contains valid JSON
      const jsonMatch = llmResponse.match(/\{[\s\S]*\}/);
      if (jsonMatch) {
        const parsed = JSON.parse(jsonMatch[0]);
        return this.validateAndEnhanceResponse(parsed);
      }
    } catch (error) {
      console.warn('Failed to extract JSON from response:', error);
    }

    // If JSON parsing fails, try to extract summary from text
    console.warn('Ollama did not return valid JSON. Attempting to extract summary from text response.');
    
    // Try to extract a meaningful summary from text
    const lines = llmResponse.split('\n').filter(line => line.trim().length > 0);
    const firstMeaningfulLine = lines.find(line => 
      line.trim().length > 20 && 
      !line.includes('```') && 
      !line.includes('{') && 
      !line.includes('}')
    );

    return {
      summary: firstMeaningfulLine || llmResponse.substring(0, 200) || 'Review response received but could not be parsed as JSON',
      findings: this.extractFindingsFromText(llmResponse),
      effortEstimation: this.extractEffortFromText(llmResponse)
    };
  }

  /**
   * Validate and enhance parsed JSON response
   */
  private validateAndEnhanceResponse(parsed: any): any {
    // Ensure required fields exist
    const response: any = {
      summary: parsed.summary || 'Code review completed',
      findings: Array.isArray(parsed.findings) ? parsed.findings : [],
      effortEstimation: parsed.effort_estimation || parsed.effortEstimation || 'N/A'
    };

    // Validate findings structure
    response.findings = response.findings.map((finding: any) => ({
      line: finding.line || null,
      type: finding.type || 'style',
      message: finding.message || finding.issue || 'Issue found',
      suggestion: finding.suggestion || finding.fix || 'Consider reviewing this'
    }));

    return response;
  }

  /**
   * Try to extract findings from text response (fallback)
   */
  private extractFindingsFromText(text: string): any[] {
    const findings: any[] = [];
    
    // Look for patterns like "Line X:", "Issue:", "Problem:", etc.
    const lines = text.split('\n');
    let currentFinding: any = null;

    for (const line of lines) {
      const lineMatch = line.match(/line\s+(\d+)/i);
      if (lineMatch) {
        if (currentFinding) {
          findings.push(currentFinding);
        }
        currentFinding = {
          line: parseInt(lineMatch[1]),
          type: 'style',
          message: line.trim(),
          suggestion: ''
        };
      } else if (currentFinding && line.trim().length > 10) {
        if (!currentFinding.message || currentFinding.message === line.trim()) {
          currentFinding.suggestion = line.trim();
        }
      }
    }

    if (currentFinding) {
      findings.push(currentFinding);
    }

    return findings.length > 0 ? findings : [];
  }

  /**
   * Try to extract effort estimation from text response (fallback)
   */
  private extractEffortFromText(text: string): string {
    // Look for patterns like "3/10", "5 out of 10", "low effort", etc.
    const effortPatterns = [
      /(\d+)\s*\/\s*10/i,
      /effort[:\s]+(\d+)/i,
      /(\d+)\s*out\s*of\s*10/i,
      /rating[:\s]+(\d+)/i
    ];

    for (const pattern of effortPatterns) {
      const match = text.match(pattern);
      if (match && match[1]) {
        return `${match[1]}/10`;
      }
    }

    // Try to infer from keywords
    const lowerText = text.toLowerCase();
    if (lowerText.includes('minor') || lowerText.includes('simple') || lowerText.includes('easy')) {
      return '2/10';
    } else if (lowerText.includes('moderate') || lowerText.includes('medium')) {
      return '5/10';
    } else if (lowerText.includes('major') || lowerText.includes('significant') || lowerText.includes('complex')) {
      return '8/10';
    }

    return 'N/A';
  }

  /**
   * Post a comment/question on a review and get AI response
   */
  postComment(projectId: number, reviewId: number, request: CommentRequest): Observable<CommentResponse> {
    return this.http.post<CommentResponse>(
      `${this.baseUrl}/${projectId}/reviews/${reviewId}/comments`,
      request,
      { headers: this.getAuthHeaders() }
    );
  }

  /**
   * Get all comments/conversation for a review
   */
  getComments(projectId: number, reviewId: number): Observable<CommentResponse[]> {
    return this.http.get<CommentResponse[]>(
      `${this.baseUrl}/${projectId}/reviews/${reviewId}/comments`,
      { headers: this.getAuthHeaders() }
    );
  }

  /**
   * Delete a guideline from a project
   */
  deleteGuideline(projectId: number, guidelineId: number): Observable<void> {
    return this.http.delete<void>(
      `${this.baseUrl}/${projectId}/guidelines/${guidelineId}`,
      { headers: this.getAuthHeaders() }
    );
  }
}

File: backend/src/main/java/com/app_template/App_Template/service/codezen/CodeZenService.java
package com.app_template.App_Template.service.codezen;

import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.app_template.App_Template.dto.CommentRequest;
import com.app_template.App_Template.dto.CommentResponse;
import com.app_template.App_Template.dto.GuidelineRequest;
import com.app_template.App_Template.dto.GuidelineResponse;
import com.app_template.App_Template.dto.ProjectRequest;
import com.app_template.App_Template.dto.ProjectResponse;
import com.app_template.App_Template.dto.ReviewRequest;
import com.app_template.App_Template.dto.ReviewResponse;
import com.app_template.App_Template.entity.CustomGuideline;
import com.app_template.App_Template.entity.Project;
import com.app_template.App_Template.entity.Review;
import com.app_template.App_Template.entity.ReviewComment;
import com.app_template.App_Template.entity.User;
import com.app_template.App_Template.repository.CustomGuidelineRepository;
import com.app_template.App_Template.repository.ProjectRepository;
import com.app_template.App_Template.repository.ReviewCommentRepository;
import com.app_template.App_Template.repository.ReviewRepository;
import com.app_template.App_Template.service.ollama.OllamaService;

import jakarta.persistence.EntityNotFoundException;
import lombok.RequiredArgsConstructor;
import reactor.core.publisher.Mono;

/**
 * CodeZenService - Main business logic for code review functionality.
 * Handles project management, code reviews, and custom guidelines.
 */
@Service
@RequiredArgsConstructor
public class CodeZenService {

    private final ProjectRepository projectRepository;
    private final ReviewRepository reviewRepository;
    private final CustomGuidelineRepository guidelineRepository;
    private final ReviewCommentRepository commentRepository;
    private final OllamaService ollamaService;

    /**
     * Create a new project for the authenticated user.
     */
    @Transactional
    public ProjectResponse createProject(ProjectRequest request, User user) {
        Project project = Project.builder()
                .name(request.getName())
                .language(request.getLanguage())
                .owner(user)
                .createdAt(LocalDateTime.now())
                .build();

        Project saved = projectRepository.save(project);
        return mapToProjectResponse(saved);
    }

    /**
     * Get all projects for the authenticated user.
     */
    @Transactional(readOnly = true)
    public List<ProjectResponse> getAllProjects(User user) {
        List<Project> projects = projectRepository.findByOwnerOrderByCreatedAtDesc(user);
        return projects.stream()
                .map(this::mapToProjectResponse)
                .collect(Collectors.toList());
    }

    /**
     * Get a specific project by ID (must belong to user).
     */
    @Transactional(readOnly = true)
    public ProjectResponse getProject(Long id, User user) {
        Project project = projectRepository.findByIdAndOwner(id, user)
                .orElseThrow(() -> new EntityNotFoundException("Project not found"));
        return mapToProjectResponse(project);
    }

    /**
     * Delete a project (must belong to user).
     */
    @Transactional
    public void deleteProject(Long id, User user) {
        if (!projectRepository.existsByIdAndOwner(id, user)) {
            throw new EntityNotFoundException("Project not found");
        }
        projectRepository.deleteById(id);
    }

    /**
     * Submit code for review and get AI feedback from Ollama.
     */
    @Transactional
    public Mono<ReviewResponse> createReview(Long projectId, ReviewRequest request, User user) {
        Project project = projectRepository.findByIdAndOwner(projectId, user)
                .orElseThrow(() -> new EntityNotFoundException("Project not found"));

        // Create review entity first
        Review review = Review.builder()
                .codeSnapshot(request.getCode())
                .timestamp(LocalDateTime.now())
                .project(project)
                .user(user)
                .build();

        Review savedReview = reviewRepository.save(review);

        // Call Ollama asynchronously
        return ollamaService.reviewCode(request.getCode(), project)
                .map(response -> {
                    // Parse effort estimation from response (if possible)
                    String effort = extractEffortEstimation(response);

                    // Update review with response
                    savedReview.setLlmResponse(response);
                    savedReview.setEffortEstimation(effort);
                    reviewRepository.save(savedReview);

                    return mapToReviewResponse(savedReview);
                });
    }

    /**
     * Get all reviews for a project.
     */
    @Transactional(readOnly = true)
    public List<ReviewResponse> getReviews(Long projectId, User user) {
        Project project = projectRepository.findByIdAndOwner(projectId, user)
                .orElseThrow(() -> new EntityNotFoundException("Project not found"));

        List<Review> reviews = reviewRepository.findByProjectOrderByTimestampDesc(project);
        return reviews.stream()
                .map(this::mapToReviewResponse)
                .collect(Collectors.toList());
    }

    /**
     * Get a specific review by ID.
     */
    @Transactional(readOnly = true)
    public ReviewResponse getReview(Long projectId, Long reviewId, User user) {
        Project project = projectRepository.findByIdAndOwner(projectId, user)
                .orElseThrow(() -> new EntityNotFoundException("Project not found"));

        Review review = reviewRepository.findByIdAndProject(reviewId, project)
                .orElseThrow(() -> new EntityNotFoundException("Review not found"));

        return mapToReviewResponse(review);
    }

    /**
     * Add a custom guideline to a project.
     */
    @Transactional
    public GuidelineResponse addGuideline(Long projectId, GuidelineRequest request, User user) {
        Project project = projectRepository.findByIdAndOwner(projectId, user)
                .orElseThrow(() -> new EntityNotFoundException("Project not found"));

        CustomGuideline guideline = CustomGuideline.builder()
                .ruleText(request.getRuleText())
                .project(project)
                .build();

        CustomGuideline saved = guidelineRepository.save(guideline);
        return mapToGuidelineResponse(saved);
    }

    /**
     * Get all guidelines for a project.
     */
    @Transactional(readOnly = true)
    public List<GuidelineResponse> getGuidelines(Long projectId, User user) {
        Project project = projectRepository.findByIdAndOwner(projectId, user)
                .orElseThrow(() -> new EntityNotFoundException("Project not found"));

        List<CustomGuideline> guidelines = guidelineRepository.findByProjectOrderById(project);
        return guidelines.stream()
                .map(this::mapToGuidelineResponse)
                .collect(Collectors.toList());
    }

    /**
     * Delete a guideline from a project (must belong to user's project).
     */
    @Transactional
    public void deleteGuideline(Long projectId, Long guidelineId, User user) {
        Project project = projectRepository.findByIdAndOwner(projectId, user)
                .orElseThrow(() -> new EntityNotFoundException("Project not found"));

        CustomGuideline guideline = guidelineRepository.findByIdAndProject(guidelineId, project)
                .orElseThrow(() -> new EntityNotFoundException("Guideline not found"));

        guidelineRepository.delete(guideline);
    }

    // Helper methods for mapping entities to DTOs

    private ProjectResponse mapToProjectResponse(Project project) {
        return ProjectResponse.builder()
                .id(project.getId())
                .name(project.getName())
                .language(project.getLanguage())
                .createdAt(project.getCreatedAt())
                .ownerId(project.getOwner().getId())
                .ownerEmail(project.getOwner().getEmail())
                .build();
    }

    private ReviewResponse mapToReviewResponse(Review review) {
        return ReviewResponse.builder()
                .id(review.getId())
                .timestamp(review.getTimestamp())
                .codeSnapshot(review.getCodeSnapshot())
                .llmResponse(review.getLlmResponse())
                .effortEstimation(review.getEffortEstimation())
                .projectId(review.getProject().getId())
                .userId(review.getUser().getId())
                .build();
    }

    private GuidelineResponse mapToGuidelineResponse(CustomGuideline guideline) {
        return GuidelineResponse.builder()
                .id(guideline.getId())
                .ruleText(guideline.getRuleText())
                .projectId(guideline.getProject().getId())
                .build();
    }

    /**
     * Extract effort estimation from LLM response (if present).
     * Handles both JSON and text responses.
     */
    private String extractEffortEstimation(String response) {
        if (response == null || response.trim().isEmpty()) {
            return null;
        }

        try {
            // Try JSON extraction first
            if (response.contains("\"effort_estimation\"")) {
                // Extract from JSON: "effort_estimation": "X/10"
                int start = response.indexOf("\"effort_estimation\"") + 20;
                int quoteStart = response.indexOf("\"", start);
                if (quoteStart > start) {
                    int quoteEnd = response.indexOf("\"", quoteStart + 1);
                    if (quoteEnd > quoteStart) {
                        String value = response.substring(quoteStart + 1, quoteEnd);
                        if (!value.trim().isEmpty()) {
                            return value;
                        }
                    }
                }
            }

            // Try alternative JSON key
            if (response.contains("\"effortEstimation\"")) {
                int start = response.indexOf("\"effortEstimation\"") + 18;
                int quoteStart = response.indexOf("\"", start);
                if (quoteStart > start) {
                    int quoteEnd = response.indexOf("\"", quoteStart + 1);
                    if (quoteEnd > quoteStart) {
                        String value = response.substring(quoteStart + 1, quoteEnd);
                        if (!value.trim().isEmpty()) {
                            return value;
                        }
                    }
                }
            }

            // Try text pattern extraction (fallback)
            // Look for patterns like "3/10", "5 out of 10", "effort: 4", etc.
            String lowerResponse = response.toLowerCase();
            
            // Pattern: "X/10" or "X out of 10"
            java.util.regex.Pattern pattern = java.util.regex.Pattern.compile("(\\d+)\\s*(?:/|out\\s+of)\\s*10", java.util.regex.Pattern.CASE_INSENSITIVE);
            java.util.regex.Matcher matcher = pattern.matcher(response);
            if (matcher.find()) {
                String effort = matcher.group(1) + "/10";
                return effort;
            }

            // Pattern: "effort: X" or "effort rating: X"
            pattern = java.util.regex.Pattern.compile("effort[\\s:]+(\\d+)", java.util.regex.Pattern.CASE_INSENSITIVE);
            matcher = pattern.matcher(response);
            if (matcher.find()) {
                String effort = matcher.group(1) + "/10";
                return effort;
            }

            // Infer from keywords
            if (lowerResponse.contains("minor") || lowerResponse.contains("simple") || lowerResponse.contains("easy")) {
                return "2/10";
            } else if (lowerResponse.contains("moderate") || lowerResponse.contains("medium")) {
                return "5/10";
            } else if (lowerResponse.contains("major") || lowerResponse.contains("significant") || lowerResponse.contains("complex")) {
                return "8/10";
            }

        } catch (Exception e) {
            // Ignore parsing errors
        }
        return null;
    }

    /**
     * Post a comment/question on a review and get AI response.
     */
    @Transactional
    public Mono<CommentResponse> postComment(Long projectId, Long reviewId, CommentRequest request, User user) {
        Project project = projectRepository.findByIdAndOwner(projectId, user)
                .orElseThrow(() -> new EntityNotFoundException("Project not found"));

        Review review = reviewRepository.findByIdAndProject(reviewId, project)
                .orElseThrow(() -> new EntityNotFoundException("Review not found"));

        // Save user's question
        ReviewComment userComment = ReviewComment.builder()
                .review(review)
                .user(user)
                .message(request.getMessage())
                .role("USER")
                .timestamp(LocalDateTime.now())
                .build();

        ReviewComment savedUserComment = commentRepository.save(userComment);

        // Get conversation history
        List<ReviewComment> history = commentRepository.findByReviewOrderByTimestampAsc(review);

        // Get AI response
        return ollamaService.answerReviewQuestion(request.getMessage(), review, history)
                .map(aiResponse -> {
                    // Save AI response
                    ReviewComment aiComment = ReviewComment.builder()
                            .review(review)
                            .user(user)
                            .message(aiResponse)
                            .role("AI")
                            .timestamp(LocalDateTime.now())
                            .build();

                    ReviewComment savedAiComment = commentRepository.save(aiComment);
                    return mapToCommentResponse(savedAiComment);
                });
    }

    /**
     * Get all comments/conversation for a review.
     */
    @Transactional(readOnly = true)
    public List<CommentResponse> getComments(Long projectId, Long reviewId, User user) {
        Project project = projectRepository.findByIdAndOwner(projectId, user)
                .orElseThrow(() -> new EntityNotFoundException("Project not found"));

        Review review = reviewRepository.findByIdAndProject(reviewId, project)
                .orElseThrow(() -> new EntityNotFoundException("Review not found"));

        List<ReviewComment> comments = commentRepository.findByReviewOrderByTimestampAsc(review);
        return comments.stream()
                .map(this::mapToCommentResponse)
                .collect(Collectors.toList());
    }

    /**
     * Map ReviewComment entity to CommentResponse DTO.
     */
    private CommentResponse mapToCommentResponse(ReviewComment comment) {
        return CommentResponse.builder()
                .id(comment.getId())
                .message(comment.getMessage())
                .role(comment.getRole())
                .timestamp(comment.getTimestamp())
                .reviewId(comment.getReview().getId())
                .userId(comment.getUser().getId())
                .build();
    }
}